pep-0484.txt

```diff
@@ -276,6 +276,33 @@ implemented by deferring to ``isinstance(x, collections.abc.Callable)``.
 However, ``isinstance(x, typing.Callable[...])`` is not supported.
 
 
+Decorators
+----------
+
+Decorators can modify the types of the functions or classes they
+decorate. To annotate a decorated declaration with the eventual
+decorated type, put a type comment on the line with the decorator::
+
+  from typing import ContextManager, Generator
+  from contextlib import contextmanager
+
+  class DatabaseSession: ...
+
+  @contextmanager  # type: Callable[[str], ContextManager[DatabaseSession]]
```

Comment:
Huh.  I can imagine a new decorator that works like so:

```python

@cast_decorator(Callable[[str], ContextManager[DatabaseSession]])
@contextmanager
def session(...): ...
```

That would have runtime inspectability.  Is it clunkier? I haven't let it bounce around in my head enough to tell.

Is that the general idea you're getting at?